Fix minor bugs due to AST and compiler changes

wazizian opened the pull request:
Hi !
First, thank you for creating this project, it's an interesting use of Cassette.jl! I was very curious about it so I tried to test it on some of my code. For this, I had to make some minor modifications to accommodate changes in the AST (now all the tests pass on Julia 1.6.3).
Maybe this could be useful to other people who happen to stumble here :-)
Bests,

wazizian committed:
Fix minor bugs due to AST and compiler changes

oxinabox commented on Project.toml:
```diff
@@ -12,13 +12,7 @@ OrderedCollections = "bac558e1-5e72-5ebc-8fee-abe8a469f55d"
 Revise = "295af30f-e4ad-537b-8983-00126c2a3abe"
 
 [compat]
-Cassette = "0.2.2"
-CodeTracking = "0.5.0"
-OrderedCollections = "1.1"
-Revise = "2.1.3"
-TimerOutputs = "~0.5.0"
-julia = "1.1"
```
we need these still, but might need different versions

wazizian replied:
Fixed in a new commit below. In particular, I now require Julia@1.6 since it's when changes in the AST occur, and Revise@3.1.5 to fix [this issue](https://github.com/timholy/Revise.jl/issues/569). I also updated Cassette and CodeTracking to accommodate new major versions.

wazizian replied:
(I tested the package with both the minimal versions in the new project file and the newest versions on Julia 1.6.3)

oxinabox commented on src/inner_repl.jl:
```diff
@@ -51,8 +51,9 @@ function run_repl(name2var, eval_module)
         code_ast = get_user_input()
-        if haskey(actions, code_ast)
-            return code_ast # Send the codeword back
+        # TODO: use REPLMaker instead of this hack ?
```
This should be in an issue
```suggestion
```

wazizian replied:
(Removed in commit below)

oxinabox commented on src/inner_repl.jl:
```diff
@@ -51,8 +51,9 @@ function run_repl(name2var, eval_module)
-            return code_ast # Send the codeword back
+        # TODO: use REPLMaker instead of this hack ?
+        if length(code_ast.args) >= 2 && haskey(actions, code_ast.args[2])
+            return code_ast.args[2]
```
Comment was deleted by mistake
```suggestion
            return code_ast.args[2]  # Send the codeword back
```

wazizian replied:
(Restored in commit below)

oxinabox commented on src/pass.jl:
```diff
@@ -160,7 +160,9 @@ it is the main method of this file, and calls all the ones defined earlier.
             # We instrument every new line, before the line.
             # So the first statement of the line is the one we replace
-            stmt isa Expr || return nothing
+            stmt isa Expr ||
+            stmt isa Core.ReturnNode ||
+            return nothing
```
Style
```suggestion
            stmt isa Expr || stmt isa Core.ReturnNode || return nothing
```

wazizian replied:
(Implemented in commit below)

oxinabox commented on src/inner_repl.jl:
```diff
@@ -51,8 +51,9 @@ function run_repl(name2var, eval_module)
-        if haskey(actions, code_ast)
-            return code_ast # Send the codeword back
+        # TODO: use REPLMaker instead of this hack ?
+        if length(code_ast.args) >= 2 && haskey(actions, code_ast.args[2])
```
I do not understand why this had to change.
Can you illuminate?

wazizian replied:
Julia parser seems to wrap toplevel expressions in a `toplevel` node. For example,
```
julia> Base.parse_input_line("CC")
:($(Expr(:toplevel, :(#= none:1 =#), :CC)))
```
I could not find this in the doc but it is mentionned in [this thread](https://discourse.julialang.org/t/toplevel-expressions/6153/3) and is implemented by [this block code in the parser](https://github.com/JuliaLang/julia/blob/83d7ebaca528904e7bcfc7f032f19d44fc1acd49/src/julia-parser.scm#L717).

oxinabox replied:
Perhaps clearer then:
```suggestion
        if Meta.isexpr(code_ast, :toplevel, 2) && haskey(actions, code_ast.args[2])
```

oxinabox reviewed (commented):
Awesome. Thanks!

Yeah, I don't know how useful MRH  is because it has huge compile times.
but it is, I think, pretty cool

wazizian committed:
Update minimal requirements

- Julia@1.6 is needed since the AST changes
- Revise@3.1.5 is required (see #569 at Revise.jl)

wazizian committed:
Implement style changes

wazizian commented:
Thank you for the feedback, I'm still pretty new to Julia.
Let me know if there's something else I should do :-)

wazizian committed:
Clarify parse of input line

Co-authored-by: Lyndon White <oxinabox@ucc.asn.au>

wazizian commented:
Thank you again for your feedback & time :-)
(Tests still pass)

oxinabox commented:
thanks!
